packages/doenetml-worker-rust/lib-doenetml-core/src/core/state_var_interfaces/text_state_var_interfaces.rs

```diff
@@ -0,0 +1,197 @@
+#[derive(Debug, Default)]
+pub struct GeneralStringStateVarInterface {
+    string_dependency_values: Vec<StateVarReadOnlyViewTyped<String>>,
+    from_single_essential: bool,
+}
+
+impl StateVarInterface<String> for GeneralStringStateVarInterface {
+    fn return_dependency_instructions(
+        &self,
+        extend_source: Option<&ExtendSource>,
+        parameters: &StateVarParameters,
+    ) -> Vec<DependencyInstruction> {
+        let mut dep_instructs: Vec<DependencyInstruction> = Vec::with_capacity(2);
+
+        if parameters.create_dependency_from_extend_source {
+            if let Some(dep_inst) =
+                create_dependency_instruction_from_extend_source(extend_source, parameters)
+            {
+                dep_instructs.push(dep_inst)
+            }
+        }
+
+        if let Some(dependency_instruction) = &parameters.dependency_instruction_hint {
+            dep_instructs.push(dependency_instruction.clone());
+        }
+
+        dep_instructs
+    }
+
+    fn save_dependencies_for_value_calculation(&mut self, dependencies: &Vec<Vec<Dependency>>) {
+        let num_dependencies = dependencies.iter().fold(0, |a, c| a + c.len());
+
+        let mut string_vals = Vec::with_capacity(num_dependencies);
+
+        for instruction in dependencies.iter() {
+            for Dependency {
+                value: dep_value, ..
+            } in instruction.iter()
+            {
+                if let StateVarReadOnlyView::String(dep_string_value) = dep_value {
+                    string_vals.push(dep_string_value.create_new_read_only_view())
+                } else {
+                    panic!("Got a non-string value for a dependency for a GeneralStringStateVarInterface");
+                }
+            }
+        }
+
+        self.string_dependency_values = string_vals;
+
+        if num_dependencies == 1 {
+            if let DependencySource::Essential { .. } = dependencies[0][0].source {
+                self.from_single_essential = true;
+            }
+        }
+    }
+
+    fn calculate_state_var_from_dependencies_and_mark_fresh(
+        &self,
+        state_var: &StateVarMutableViewTyped<String>,
+    ) {
+        if self.from_single_essential {
+            // if we are basing it on a single essential variable,
+            // then we propagate used_default as well as the value.
+            state_var.set_value_and_used_default(
+                self.string_dependency_values[0].get_fresh_value().clone(),
+                self.string_dependency_values[0].get_used_default(),
+            );
+        } else {
+            // TODO: can we implement this without cloning the inner value?
+            let value: String = self
+                .string_dependency_values
+                .iter()
+                .map(|v| v.get_fresh_value().clone())
+                .collect();
+
+            state_var.set_value(value);
+        }
+    }
+
+    fn request_dependencies_to_update_value(
```

Comment:
What is `_value` adding to this name?

Reply:
I felt like the name `request_dependencies_to_update` made it sound like you are updating the actual dependencies (i.e., changing the dependency graph). The name was trying to convey that we are just requesting to change the value of the dependency (i.e., `dependency.value`). We could simplify it to `request_dependencies_to_update` or other options such as: `request_new_dependency_values`, `request_updated_dependency_values`, `request_dependency_updates`.

Reply:
I like `request_dependency_updates`

Reply:
OK. I'm working on this function now, so can change the name to `request_dependency_updates`.